### PR TITLE
Use zero pages list before virtualalloc

### PR DESCRIPTION
**Pull request by leirocks**
Before allocating new pages, check if there's zero pages or pending zero pages from background sweeping, if yes use them instead of allocating new pages

### CODE REVIEW

**Comment by leirocks** [comment]
@curtisman @digitalinfinity can you review please?

File: lib/Common/Memory/PageAllocator.cpp [hunk 1]
```diff
@@ -749,70 +749,113 @@ PageAllocatorBase<T>::AddPageSegment(DListBase<PageSegmentBase<T>>& segmentList)
...
+                this->FillAllocPages(pages, pageCount);
+                return pages;
             }
-            this->FillAllocPages(pages, pageCount);
-            return (char *)pages;
         }
     }
-
-    return nullptr;
+
```
Comment (thefourtheye) [inline lib/Common/Memory/PageAllocator.cpp]: Unnecessary whitespace change.

File: lib/Common/Memory/RecyclerPageAllocator.cpp [hunk 2]
```diff
@@ -25,6 +25,24 @@ bool RecyclerPageAllocator::IsMemProtectMode()
     return recycler->IsMemProtectMode();
 }
 
+char * RecyclerPageAllocator::TryAllocFromZeroPages(uint pageCount, PageSegmentBase<VirtualAllocWrapper> ** pageSegment, PageHeapMode pageHeapFlags)
```
Comment (thefourtheye) [inline lib/Common/Memory/RecyclerPageAllocator.cpp]: inconsistent pointer syntax (check line 30 in the same file)

File: lib/Common/Memory/PageAllocator.cpp [hunk 3]
```diff
@@ -749,70 +749,113 @@ PageAllocatorBase<T>::AddPageSegment(DListBase<PageSegmentBase<T>>& segmentList)
...
-            }
+        return TryAllocFromZeroPagesList(pageCount, pageSegment, backgroundPageQueue->freePageList);
+    }
+    return nullptr;
+}
 
-            this->freePageCount -= pageCount;
-            *pageSegment = freeSegment;
+template<typename T>
+template <bool notPageAligned>
```
Comment (thefourtheye) [inline lib/Common/Memory/PageAllocator.cpp]: inconsistent with the previous line

File: lib/Common/Memory/RecyclerPageAllocator.cpp [hunk 4]
```diff
@@ -25,6 +25,24 @@ bool RecyclerPageAllocator::IsMemProtectMode()
...
+
+    pages = TryAllocFromZeroPagesList(pageCount, pageSegment, this->zeroPageQueue.pendingZeroPageList);
+    if (pages != nullptr)
+    {
+        // zero out
+        memset(pages, 0, pageCount*AutoSystemInfo::PageSize);
+        return pages;
+    }
+
+    pages = TryAllocFromZeroPagesList(pageCount, pageSegment, this->zeroPageQueue.freePageList);
```
Comment (curtisman) [inline lib/Common/Memory/RecyclerPageAllocator.cpp]: We should try the freePageList first before the pendingZeroPageList,as it is cheaper without the need to zero the pages first.

File: lib/Common/Memory/PageAllocator.cpp [hunk 5]
```diff
@@ -749,70 +749,113 @@ PageAllocatorBase<T>::AddPageSegment(DListBase<PageSegmentBase<T>>& segmentList)
...
+            break;
+        }
+        else if (freePage->pageCount > pageCount)
+        {
+            *pageSegment = freePage->segment;
+            freePage->pageCount -= pageCount;
+            pages = (char *)freePage + freePage->pageCount * AutoSystemInfo::PageSize;
+            this->FillAllocPages(pages, pageCount);
+
+            ::InterlockedPushEntrySList(&backgroundPageQueue->freePageList, freePage);
```
Comment (curtisman) [inline lib/Common/Memory/PageAllocator.cpp]: If this is the pendingZeroPageList, you need to zero out all the page before putting it back to the freePageList.
You can't put it back to the pendingZeroPageList as it may be missed by the background GC thread's BackgroundZeroQueuedPages calls.

File: lib/Common/Memory/RecyclerPageAllocator.cpp (hunk 4 above)
Comment (leirocks) [inline lib/Common/Memory/RecyclerPageAllocator.cpp]: yeah, I hesitated here. it was freePages first. Then I thought pendingZeroPage might have bigger chance to hit. But I think you are right, checking existence of freePages cost should be very low comparing to zero out total 4 pages. let me change back

File: lib/Common/Memory/PageAllocator.cpp [hunk 6]
```diff
@@ -749,70 +749,113 @@ PageAllocatorBase<T>::AddPageSegment(DListBase<PageSegmentBase<T>>& segmentList)
...
+        }
+
+        freePage = (FreePageEntry *)::InterlockedPopEntrySList(&zeroPagesList);
+    }
+
+    while (localList != nullptr)
     {
-        return nullptr;
+        FreePageEntry* next = (FreePageEntry*)localList->Next;
+        ::InterlockedPushEntrySList(&zeroPagesList, localList);
```
Comment (curtisman) [inline lib/Common/Memory/PageAllocator.cpp]: Same comment as before. You can't put it back to the pendingZeroPageList as it may be missed by the background GC thread's BackgroundZeroQueuedPages calls.

Alternatively, you will have to add a "last chance" ZeroQueuedPages on the main thread before we to ConcurrentTransferSwept.

File: lib/Common/Memory/PageAllocator.cpp (hunk 5 above)
Comment (leirocks) [inline lib/Common/Memory/PageAllocator.cpp]: hmm.. then I need to zero out all checked out pendingZero pages in below while loop..

File: lib/Common/Memory/PageAllocator.cpp [hunk 7]
```diff
@@ -749,70 +749,147 @@ PageAllocatorBase<T>::AddPageSegment(DListBase<PageSegmentBase<T>>& segmentList)
...
+                this->FillAllocPages(pages, pageCount);
+                break;
+            }
+        }
+    }
+
+    if (localList != nullptr)
+    {
+        uint newFreePages = 0;
+        while (true)
```
Comment (curtisman) [inline lib/Common/Memory/PageAllocator.cpp]: Why not "do { } while (localList != nullptr)"?

Comment (leirocks) [inline lib/Common/Memory/PageAllocator.cpp]: hmm, localList can be null. but I can  remove the inner if check with while(localList!=nullptr){}

File: lib/Common/Memory/RecyclerPageAllocator.h [hunk 8]
```diff
@@ -40,5 +40,6 @@ class RecyclerPageAllocator : public IdleDecommitPageAllocator
     Recycler* recycler;
 
     bool IsMemProtectMode();
+    virtual char * TryAllocFromZeroPages(uint pageCount, PageSegmentBase<VirtualAllocWrapper> ** pageSegment, PageHeapMode pageHeapFlags) override;
```
Comment (curtisman) [inline lib/Common/Memory/RecyclerPageAllocator.h]: Can we avoid the virtual here? 
I think you can just check for HasZeroPageQueue()?

Comment (leirocks) [inline lib/Common/Memory/RecyclerPageAllocator.h]: OK

Comment (leirocks) [inline lib/Common/Memory/RecyclerPageAllocator.h]: also looks there's a bug that I checked hasZeroQueuedPages before allocating from zero page list. should check for that only for pending zero page list allocating

File: lib/Common/Memory/PageAllocator.cpp [hunk 9]
```diff
@@ -749,70 +749,141 @@ PageAllocatorBase<T>::AddPageSegment(DListBase<PageSegmentBase<T>>& segmentList)
...
+PageAllocatorBase<T>::TryAllocFreePages(uint pageCount, PageSegmentBase<T> ** pageSegment, PageHeapMode pageHeapFlags)
+{
+    Assert(!HasMultiThreadAccess());
+    char* pages = nullptr;
+
+    if (this->freePageCount >= pageCount)
     {
-        FreePageEntry * freePage = (FreePageEntry *)::InterlockedPopEntrySList(&backgroundPageQueue->freePageList);
-        if (freePage != nullptr)
+        FAULTINJECT_MEMORY_NOTHROW(this->debugName, pageCount*AutoSystemInfo::PageSize);
```
Comment (curtisman) [inline lib/Common/Memory/PageAllocator.cpp]: Should this be outside of the "if"? Or we need to add it in TryAllocFromZeroPagesList?

Comment (leirocks) [inline lib/Common/Memory/PageAllocator.cpp]: hmm. yes

**Comment by curtisman** [comment]
LGTM
